xpdan/search.py

```diff
@@ -0,0 +1,186 @@
+from pyxdameraulevenshtein import \
+    normalized_damerau_levenshtein_distance as ndld
+import numpy as np
+from databroker.broker import _munge_time
+from pprint import pprint
+import pytz
+
+
+def getFromDict(dataDict, mapList):
+    """ Get a value from a nested dictionary, given a list of keys
+
+    Parameters
+    ----------
+    dataDict: dict
+        The dictionary to be queried
+    mapList: list of str
+        A list of strings, each string is one level lower than the previous
+
+    Returns
+    -------
+    object:
+        The the value from the dict
+
+    """
+    for k in mapList:
+        dataDict = dataDict[k]
+    return dataDict
+
+
+def nested_dict_values(d):
+    """Yield all string values inside a nested dictionary
+
+    Parameters
+    ----------
+    d: dict
+        The dictionary to be unpacked
+
+    Yields
+    -------
+    str:
+        The string value inside the dictionary
+
+    """
+    for v in d.values():
+        if isinstance(v, dict):
+            yield from nested_dict_values(v)
+        else:
+            if isinstance(v, str):
+                yield v
+            else:
+                yield None
+
+
+def fuzzy_search(db, keys, search_string, size=100):
+    """Fuzzy search a databroker for given keys
+
+    Parameters
+    ----------
+    db: databroker.DataBroker instance
+        The databroker to be searched
+    keys: list of str
+        The list of strings to be accessed
+    search_string: str
+        The string to be searched for
+    size: int or 'all', optional
+        The number of results to be returned, if 'all' all are returned.
+         Defaults to 100 results
+
+    Returns
+    -------
+    list:
+        A list
+
+    """
+    if isinstance(keys, list):
+        scores = []
+        for h in db():
+            scores.append(ndld(getFromDict(h['start'], keys), search_string))
+        scores = [ndld(getFromDict(h['start'], keys), search_string) for h in
+                  db()]
+    else:
+        scores = [ndld(h['start'][keys], search_string) for h in db()]
+    zipped = zip(scores, db())
```

Comment:
there is a race condition here if an event gets added to the db between the first call and now.  It is better to grab the uid on the first pass.  In the next line you will pull everything into memory anyway so if you are going to do it this way, just make a list of header at the start.  

A better option is to keep a fixed size heap (see the `heapq` module) of the number you want to return, then do a single pass through the headers.  At each header use `heapq.heappushpop` https://docs.python.org/3.5/library/heapq.html#heapq.heappushpop to only keep the currntly seen `N` best.  As the last step, sort and return the relevant headers. 

I suggest using a heap element that looks like `(ndl(), header)`.